### PR TITLE
Provide somewhat sensible defaults for all hyperparameters

### LINKED ISSUE #307: Provide sensible defaults for all(most?) hyperparameters
To streamline the (common) user experience, I think we should provide defaults for most hyperparameters of the library. The defaults should be taken from the CGP literature, possibly with some exceptions, e.g., ~10 columns, rather than thousands.

I believe this would significantly improve the user experience by not having to worry about many hyperparameter settings, but rather starting from a reasonable default, figuring out which should be changed.

Thoughts @mschmidt87 @HenrikMettler ?

### PR DESCRIPTION
**Pull request by jakobj**
as the title suggests

we could also simplify the minimal example a lot with these new defaults. what do you think?

closes #307

### CODE REVIEW

File: cgp/genome.py [hunk 1]
```diff
@@ -26,30 +26,30 @@ class Genome:
...
-        n_inputs: int,
-        n_outputs: int,
-        n_columns: int,
-        n_rows: int,
-        primitives: Tuple[Type[Node], ...],
+        n_inputs: int = 1,
+        n_outputs: int = 1,
+        n_columns: int = 12,
+        n_rows: int = 1,
+        primitives: Optional[Tuple[Type[Node], ...]] = None,
```
Comment (HenrikMettler) [inline cgp/genome.py:33]:
```python
primitives: Optional[Tuple[Type[Node], ...]] = (Add, Sub, Mul, ConstantFloat),
```
to be consistent with the doc string

File: cgp/genome.py [hunk 2]
```diff
@@ -76,7 +76,17 @@ def __init__(
             raise ValueError("levels_back can not be larger than n_columns")
         self._levels_back = levels_back
 
-        self._primitives = Primitives(primitives)
+        if primitives is None:
+            from .node_impl import (  # delayed to avoid circular imports
```
Comment (HenrikMettler) [inline cgp/genome.py:80]: Why would you have circular imports?

File: cgp/genome.py [hunk 3]
```diff
@@ -26,30 +26,30 @@ class Genome:
 
     def __init__(
         self,
-        n_inputs: int,
-        n_outputs: int,
-        n_columns: int,
-        n_rows: int,
-        primitives: Tuple[Type[Node], ...],
+        n_inputs: int = 1,
```
Comment (HenrikMettler) [inline cgp/genome.py:29]: Line 9 (sorry cant write there somehow): 
from .node_impl import Add, ConstantFloat, Mult, Sub

File: cgp/hl_api.py [hunk 4]
```diff
@@ -9,44 +9,43 @@
...
 def evolve(
-    pop: Population,
     objective: Callable[[IndividualBase], IndividualBase],
-    ea: MuPlusLambda,
+    pop: Population = Population(),
+    ea: MuPlusLambda = MuPlusLambda(),
     min_fitness: float = np.inf,
     termination_fitness: float = np.inf,
-    max_generations: int = np.iinfo(np.int64).max,
+    max_generations: int = 1000,
```
Comment (HenrikMettler) [inline cgp/hl_api.py:17]: Not sure this change is a good idea. If the user wants to use `max_objective_calls` this default will still be used (unless overwritten)

File: cgp/hl_api.py [hunk 5]
```diff
@@ -9,44 +9,43 @@
...
-        Minimum fitness at which the evolution is terminated
-    max_generations : int
+    termination_fitness : float, optional
+        Minimum fitness at which the evolution is terminated. Defaults to positive infinity.
+    max_generations : int, optional
         Maximum number of generations.
-        Defaults to largest representable integer.
-        Either this or `max_objective_calls` needs to be set to a value less than that.
-    max_objective_calls: int
+        Defaults to 1000.
```
Comment (HenrikMettler) [inline cgp/hl_api.py:43]: see above

File: cgp/hl_api.py [hunk 6]
```diff
@@ -55,7 +54,8 @@ def evolve(
 
     Returns
     -------
-    None
+    Population
```
Comment (HenrikMettler) [inline cgp/hl_api.py:57]: I think this is a good idea, but should we maybe put it in a separate PR?

File: cgp/population.py [hunk 7]
```diff
@@ -13,24 +13,24 @@ class Population:
...
-            Number of parent individuals.
-        seed : int
-            Seed for internal random number generator.
+        n_parents : int, optional
+            Number of parent individuals. Defaults to 4.
+        seed : int, optional
+            Seed for internal random number generator. Defaults to 1234.
         genome_params : dict
-            Parameters for the genomes of the population's individuals.
+            Parameters for the genomes of the population's individuals. Defaults to None.
```
Comment (HenrikMettler) [inline cgp/population.py:30]: Hmm might it be more user-friendly to have the defaults here instead of nested in `Genome()`?

Comment (HenrikMettler) [reply cgp/population.py:30]: At least the defaults should be documented here as well

File: examples/example_caching.py [hunk 8]
```diff
@@ -98,7 +98,7 @@ def objective(individual):
 # Finally, we call the `evolve` method to perform the evolutionary search.
 
 
-cgp.evolve(pop, objective, ea, **params["evolve_params"], print_progress=True)
+cgp.evolve(objective, pop, ea, **params["evolve_params"], print_progress=True)
```
Comment (HenrikMettler) [inline examples/example_caching.py:101]: I think we should be consistent in calling it `objective` or `obj` in all examples (but maybe this can be a separate maintenance PR)

File: test/test_utils.py [hunk 9]
```diff
@@ -66,7 +66,7 @@ def objective(ind):
         def recording_callback(pop):
             history["fitness_champion"].append(pop.champion.fitness)
 
-        cgp.evolve(pop, objective, ea, **evolve_params, callback=recording_callback)
+        cgp.evolve(objective, pop, ea, **evolve_params, callback=recording_callback)
```
Comment (HenrikMettler) [inline test/test_utils.py:69]: Same here: Name them all `obj` or `objective`

**Review by HenrikMettler — CHANGES_REQUESTED** [review]
Thanks for working on this :) Some points on parameter-values and user-friendliness.

File: cgp/genome.py (hunk 2 above)
Comment (HenrikMettler) [reply cgp/genome.py:80]: Understood, but please describe circle

File: examples/example_caching.py (hunk 8 above)
Comment (HenrikMettler) [reply examples/example_caching.py:101]: see separate issue #336

**Comment by HenrikMettler** [comment]
What a nice opportunity to update the readme ;)

File: cgp/population.py (hunk 7 above)
Comment (jakobj) [reply cgp/population.py:30]: i avoided documenting the results here but put a reference to where they are defined. if we duplicate this information one part will become outdated quickly.

**Comment by jakobj** [comment]
thanks @HenrikMettler for the review. i've addressed all your comments, rebased on master and squashed the commits. please have another look.

**Comment by jakobj** [comment]
ping @HenrikMettler

**Review by HenrikMettler — APPROVED** [review]
LGTM, thanks for taking care of it :) Please squish the `fix formatting` commit

**Comment by HenrikMettler** [comment]
ping @jakobj
